scipy/linalg/_decomp_ldl.py

```diff
@@ -0,0 +1,346 @@
+        raise ValueError('{} exited with the internal error "illegal value '
+                         'in argument number {}". See LAPACK documentation '
+                         'for the error codes.'.format(s.upper(), -info))
+
+    swap_arr, pivot_arr = _ldl_sanitize_ipiv(piv, lower=lower)
+    d, lu = _ldl_get_d_and_l(ldu, pivot_arr, lower=lower, hermitian=hermitian)
+    lu, perm = _ldl_construct_tri_factor(lu, swap_arr, pivot_arr, lower=lower)
+
+    return lu, d, perm
+
+
+def _ldl_sanitize_ipiv(a, lower=True):
+    """
+    This helper function takes the rather strangely encoded permutation array
+    returned by the LAPACK routines ?(HE/SY)TRF and converts it into
+    regularized permutation and diagonal pivot size format.
+
+    Since FORTRAN uses 1-indexing and LAPACK uses different start points for
+    upper and lower formats there are certain offsets in the indices used
+    below.
+
+    Let's assume a result where the matrix is 6x6 and there are two 2x2
+    and two 1x1 blocks reported by the routine. To ease the coding efforts,
+    we still populate a 6-sized array and fill zeros as the following ::
+
+        pivots = [2, 0, 2, 0, 1, 1]
+
+    This denotes a diagonal matrix of the form ::
+
+        [x x        ]
+        [x x        ]
+        [    x x    ]
+        [    x x    ]
+        [        x  ]
+        [          x]
+
+    In other words, we write 2 when the 2x2 block is first encountered and
+    automatically write 0 to the next entry and skip the next spin of the
+    loop. Thus, a separate counter or array appends to keep track of block
+    sizes are avoided. Zeros can be removed much easier instead.
+
+    Parameters
+    ----------
+    a : ndarray
+        The permutation array ipiv returned by LAPACK
+    lower : bool, optional
+        The switch to select whether upper or lower triangle is chosen in
+        the LAPACK call.
+
+    Returns
+    -------
+    swap_ : ndarray
+        The array that defines the row/column swap operations. For example,
+        if row two is swapped with row four, the result is [0, 3, 2, 3].
+
+    """
+    n = a.size
+    swap_ = arange(n)
+    pivots = zeros_like(swap_, dtype=int)
+    skip_2x2 = False
+
+    # Some upper/lower dependent offset values
+    # range (s)tart, r(e)nd, r(i)ncrement
+    x, y, rs, re, ri = (1, 0, 0, n, 1) if lower else (-1, -1, n-1, -1, -1)
+
+    for ind in range(rs, re, ri):
+        # If previous spin belonged already to a 2x2 block
+        if skip_2x2:
+            skip_2x2 = False
+            continue
+
+        cur_val = a[ind]
+        # do we have a 1x1 block or not?
+        if cur_val > 0:
+            if cur_val != ind+1:
+                # Index value != array value --> permutation required
+                swap_[ind] = swap_[cur_val-1]
+            pivots[ind] = 1
+        # Not.
+        elif cur_val < 0 and cur_val == a[ind+x]:
+            # first neg entry of 2x2 block identifier
+            if -cur_val != ind+2:
```

Comment:
It is taken care of by the line `x, y, rs, re, ri = (1, 0, 0, n, 1) if lower else (-1, -1, n-1, -1, -1)`. So for upper, the indices are going backwards and this is 2x2 branch so we jump over the element. It is indeed very confusing how LAPACK decided to encode this information. I hope I understand your remark correctly.

Reply:
I'm asking because on the line above, the assignment is `swap_[ind+x] = swap_[-cur_val-1]` but the if statement checks for `-cur_val != ind+2` and not for `-cur_val != ind+x+1` so it's not analogous to the 1x1 case.

Reply:
Yes, this is the 2x2 branch and for some reason this is how they decided to do it. The extra shift comes from the [LAPACK `ipiv`](http://www.netlib.org/lapack/explore-html/d3/db6/group__double_s_ycomputational_gad91bde1212277b3e909eb6af7f64858a.html#gad91bde1212277b3e909eb6af7f64858a) which really bothered me also for quite some time.

Quoting the relevant part 

>If `UPLO = 'U'` and `IPIV(k) = IPIV(k-1) < 0`, then rows and columns `k-1` and `-IPIV(k)` were interchanged and `D(k-1:k,k-1:k)` is a 2-by-2 diagonal block.  If `UPLO = 'L'` and `IPIV(k) =IPIV(k+1) < 0`, then rows and columns `k+1` and `-IPIV(k)` were interchanged and `D(k:k+1,k:k+1)` is a 2-by-2 diagonal block.

so not swap row k and IPIV(k) as was for 1x1 but the previous(next) one and IPIV(k).